### DISCUSSION TITLE
UDIM Visualisation - UV Editing

### DISCUSSION THREAD

**Discussion by Kieruss** [StjerneIdioten/I3D-Blender-Addon#123] [Ideas & Feature Requests]
Problem: It's difficult to see where each segment begins/ends, and what those segments are (I often select faces via the UV Editor and it can be hard to tell if I've accidentally grabbed some from the next quadrant)

Solution: A grid overlay, showing the divisions
There is a built-in solution, but it only shows when no image is loaded... 
![image](https://user-images.githubusercontent.com/24835155/111585394-5b277200-8813-11eb-84c7-9aab73c27203.png)
Just the borders, and not the sub-divisions, would be fine, but have it as a toggle-able overlay

Bonus: Include on overlay of names and/or previews for each grid, like the "Pick UDIMs" map

**Comment by kbrandwijk** [comment]
Personally, I think this one borders on being a Blender feature request, and I don't know if this specific addon should address this.

**Reply by Kieruss** [reply]
The grid being visible while an image is loaded, and including negative coords, sure. However, the labelling (via text and/or example icon) isn't going to be a general Blender feature. I'm not sure how difficult it would be to draw over the image, but the addon already knows the grid sizes, locations, and names specific to what we're doing.

A simple way to implement the grid overlay would be with the annotations system. That would also provide an existing way to toggle it on/off and change the colour. Here's a really bad example of that :)
![image](https://user-images.githubusercontent.com/24835155/111718607-5a8fe980-88ae-11eb-8e51-4227ba5daba8.png)
That should (?) be the easiest implementation method, since you just need to script the drawing of a grid, and the base feature handles panning, zooming, and visibility.

Ideally, it would be good to have each grid area highlight, as you mouse over, with a border, example, and name of where you're at
![image](https://user-images.githubusercontent.com/24835155/111719044-2bc64300-88af-11eb-9bb2-591e0afbdaa5.png)

**Reply by StjerneIdioten** [reply]
> Personally, I think this one borders on being a Blender feature request, and I don't know if this specific addon should address this.

I've been eyeing [this](https://developer.blender.org/T72390) over at the blender dev hub 😉  It seems to add a lot more stuff for UDIM's, might be something I can utilize eventually. I am unfortunately pretty reliant on what they expose in the python API and so far it hasn't been much for UDIM's, the few things that exists are badly documented at best 🤣

> The grid being visible while an image is loaded, and including negative coords, sure. However, the labelling (via text and/or example icon) isn't going to be a general Blender feature. I'm not sure how difficult it would be to draw over the image, but the addon already knows the grid sizes, locations, and names specific to what we're doing.

Sort of, it is all derived from the UV-coordinates of the individual vertices. There is actually no way for me to tell what "tile" something belongs to by itself, since imagine something being at UV-coordinate (1,1), there is 4 different tiles that vertex could belong to 🤣 Even the actual UV-islands aren't stored in any way in blender, I have to go through the entire vertex list and figure out which are connected to each other.
Unless you mean how the Giants UDIM-grid is laid out, I have that information.

> A simple way to implement the grid overlay would be with the annotations system. That would also provide an existing way to toggle it on/off and change the colour. Here's a really bad example of that :)
> That should (?) be the easiest implementation method, since you just need to script the drawing of a grid, and the base feature handles panning, zooming, and visibility.
> 
> Ideally, it would be good to have each grid area highlight, as you mouse over, with a border, example, and name of where you're at

Interesting idea, I haven't looked at how much is available in the python API when it comes to actually drawing stuff inside of the viewport/editor view. I at least know you can do a lot in the 3d-viewport, since a lot of big addons like Hardops and Boxcutter draws custom shapes. Don't know how much of that translates into the 2d-viewports.

The same goes for the annotation system, I haven't checked how much is "programmable" when it comes to that. It can probably be controller using operators.

Also, I quickly banged the UDIM-editor together in a couple of days, because I got annoyed at the other one that is available. So it could do with a refresh anyway, especially with that annoying recursion "bug" that it currently has 😛
